### ISSUE TITLE
which alphanumerical keyboard to conect to ESP32 for using getc, scanf, getline?

### ISSUE THREAD

**Issue by dsyleixa** [espressif/arduino-esp32#8250]
### Related area

ESP32 keyboard

### Hardware specification

ESP32

### Is your feature request related to a problem?

which alphanumerical keyboard to conect to ESP32 for using getc, scanf, getline?

### Describe the solution you'd like

102 keyboard usage  like with common USB keyboards, stdio/iostream/stdin compatible, directly attached to PC or RaspberryPi (no WiFi or internal BT required)

### I have checked existing list of Feature requests and the Contribution Guide

- [X] I confirm I have checked existing list of Feature requests and Contribution Guide.

**Comment by SuGlider** [comment]
@dsyleixa - I'm sorry, but I didn't understand the question. 
Do you want to use a Keyboard attached to the computer that will be used to send data to the ESP32 using UART?
Please detail it further.

**Comment by dsyleixa** [comment]
no, a sort of common 102 keyboard directly attached to the ESP32, for stand-allone usage.
Sorry for my poor English.

**Comment by SuGlider** [comment]
ESP32 has no  USB HOST, thus, the only way to attach a keyboard directly to an ESP32 is by using a PS/2 keyboard that requires 2 GPIOs (DATA and CLK) plus VCC and GND.

There are a few libraries that can help with it:
https://github.com/techpaul/PS2KeyAdvanced   -- It has support to scanning most key codes.
https://github.com/michalhol/ps2kbdlib

**Comment by SuGlider** [comment]
About supporting functions ( getc, scanf, getline, including stdio/iostream/stdin), it would require creating them by scanning the KeyCodes, forming the C strings and then passing it to the standard C functions.

It would be like having a PS/2 driver in a first low level layer and then a bounding with a C level equivalent functions created for this purpose.

**Comment by dsyleixa** [comment]
as stated, directly using C/C++ getc, scanf, getline functions is crucial, at least the PS2KeyAdvanced is absolutely crappy.

**Comment by SuGlider** [comment]
I see. At this point we will need community help to work on it due to lack of time.

**Comment by dsyleixa** [comment]
perhaps possibly  also a USB-Hardware-Interface-Device, to be put in between the USB connector and the USB keyboard?

**Comment by dsyleixa** [comment]
Thanks, but unfortuantely I only use Arduino IDE 1.8.x and arduino-esp32 API. OTOH, I don't recognize how ESP32 Soft USB HID Host Library https://github.com/tobozo/ESP32-USB-Soft-Host works with getc, scanf, getline exactly via Arduino IDE.

**Comment by SuGlider** [comment]
Using Tobozo's library would need a software layer to support the standard C functions getc, scanf, getline etc.
@dsyleixa - you may try to reach @tobozo and ask for help.

@tobozo - can you add some comment to this talk?

**Comment by dsyleixa** [comment]
BTW, forgot to mention: 
I'm visually impaired so I can't solder, so it really should be plug and play...

**Comment by SuGlider** [comment]
> BTW, forgot to mention: I'm visually impaired so I can't solder, so it really should be plug and play...

For that case, maybe, it is better to use the S2 or S3. The DevKits have both USB (UART CP2102 and Native USB Host) plugs for connecting a OTG cable directly (no soldering). 

Or, maybe, using some protoboard and dupont cables?!

**Comment by SuGlider** [comment]
OK. The possible solution seems to be using your Adafruit Feather ESP32, that seems to have at aleast 20 GPIOs, to connet the SPI TFT and use 2 GPIOs for the Tobozo's Soft USB Host library. All connected using protoboard and dupont cables.

It requires some coding for the `getc(), scanf() and getline()` functions to be integrated into the pipeline of the Soft USB Host library.

**Comment by dsyleixa** [comment]
actually, tbh, I would like to connect the keyboard directly to the Feather USB port (used for Sketch upload and USB power)...
Or a BT dongle attached to it, and then using a BT keyboard...

**Comment by dsyleixa** [comment]
actually, tbh, I would like to connect the keyboard directly to the Feather USB port (used for Sketch upload and USB power)...
Or a BT dongle attached to it, and then using a BT keyboard (plus a Y-switch or hub for external USB power cable)...

**Comment by SuGlider** [comment]
> actually, tbh, I would like to connect the keyboard directly to the Feather USB port (used for Sketch upload and USB power)... Or a BT dongle attached to it, and then using a BT keyboard (plus a Y-switch or hub for external USB power cable)...

Sorry, but this is impossible given that the Feather USB port is actually an UART-USB chip. It can't work as USB HOST.

**Comment by SuGlider** [comment]
Thanks @dsyleixa. 
I'll close this issue given that it won't be necessary anymore.
